File: src/main/java/fi/quanfoxes/assembler/references/NumberReference.java
package fi.quanfoxes.assembler.references;

import fi.quanfoxes.assembler.*;

public class NumberReference extends Reference {
    private Number number;

    public NumberReference(Number number, Size size) {
        super(size);
        this.number = number;
    }

    public Number getNumberValue() {
        return number;
    }

    @Override
    public String use(Size size) {
        return String.format("%s %s", size, number.toString());
    }

    public LocationType getType() {
        return LocationType.NUMBER;
    }
}

File: Zigzag/Assembler/References/NumberReference.cs
using System.Collections.Generic;

public class NumberReference : Reference
{
	public object Value { get; private set; }

	public NumberReference(object value, Size size) : base(size)
	{
		Value = value;
	}

	public override string Use(Size size)
	{
		return string.Format("{0} {1}", size, Value.ToString().Replace(',', '.'));
	}

	public override string Use()
	{
		return Value.ToString().Replace(',', '.');
	}

	public override bool IsComplex()
	{
		return false;
	}

	public override LocationType GetType()
	{
		return LocationType.NUMBER;
	}

	public override bool Equals(object? obj)
	{
		return obj is NumberReference reference &&
			   EqualityComparer<object>.Default.Equals(Value, reference.Value);
	}
}